### PR TITLE
arch: arm: cortex_m: Move bss/noinit sections to the end of linker script

### PR DESCRIPTION
**Pull request by gramsay0**
This results in a smaller binary with `CONFIG_XIP=n` as these NOLOAD sections can be excluded

bss/noinit have been included in `CONFIG_XIP=n` binaries by default since c107827bb7456d7ead8af0b1febe932e1b2c4423. This also occurs if using iterable sections or adding code/data to ITCM/DTCM sections. Moving bss/noinit to the end of the linker file should ensure they are always excluded

Partial fix for #72307

### CODE REVIEW

**Comment by ithinuel** [comment]
@MaureenHelm 
This seems oddly related to https://github.com/zephyrproject-rtos/zephyr/pull/76651 

TBH there’s so much things going on in those files I can’t tell if something is right in all circumnstances or not.
Do we have any linker & Zephyr experts in the community who could help review this?

**Comment by gramsay0** [comment]
A similar issue was also recently fixed for x86: https://github.com/zephyrproject-rtos/zephyr/pull/73044

If you look back through the issues/PRs this has been broken and fixed many times across different architectures.

It really needs a test; a big static storage variable (say `uint8_t some_bss[KB(64)]`) and a twister or cmake post build assert on the size of the .bin file would probably be enough

**Comment by tejlmand** [comment]
> bss/noinit have been included in CONFIG_XIP=n binaries by default since https://github.com/zephyrproject-rtos/zephyr/commit/c107827bb7456d7ead8af0b1febe932e1b2c4423.

actually it was not c107827bb7456d7ead8af0b1febe932e1b2c4423 which introduced the `.last_section` for proper rom usage calculation, the referenced commit only fixed an issue whereby the location counter and actual usage could get out-of-sync.

The `.last_section` was introduced in https://github.com/zephyrproject-rtos/zephyr/commit/d85efe0b10f1f8ed4b854717f994062a4bbffe26.

> these NOLOAD sections can be excluded

they should normally be excluded from loading by the linker if the linker script is properly written.
I wonder if the `bss` section for the target on which you see the has mistakenly been placed in rom instead of ram ?

One guess could be the internal macro definition of `ROM_REGION` vs `RAM_REGION` which could be wrongly handled when `XIP=y`.

Taking a look at the issue reported to reproduce this locally before approving a  linker script adjustment, even though placing bss later should not impact rom usage on xip targets. I have a hunch that something else has been overlooked.

**Comment by gramsay0** [comment]
[heart]         Grant Ramsay reacted to your message:
________________________________

**Comment by gramsay0** [comment]
[0]     Grant Ramsay reacted to your message:
________________________________

**Comment by gramsay0** [comment]
[celebrate]     Grant Ramsay reacted to your message:
________________________________

**Comment by gramsay0** [comment]
> actually it was not [c107827](https://github.com/zephyrproject-rtos/zephyr/commit/c107827bb7456d7ead8af0b1febe932e1b2c4423) which introduced the `.last_section` for proper rom usage calculation, the referenced commit only fixed an issue whereby the location counter and actual usage could get out-of-sync.

[c107827](https://github.com/zephyrproject-rtos/zephyr/commit/c107827bb7456d7ead8af0b1febe932e1b2c4423) removed `NOLOAD` from `.last_section`, before that commit it was not included in the binary
```diff
-SECTION_PROLOGUE(.last_section,(NOLOAD),)
+SECTION_PROLOGUE(.last_section,,)
```

(P.S. sorry about the email spam, I accidentally clicked an emoji button on my mobile email client...)

**Comment by gramsay0** [comment]
I went through the ARM Cortex-M linker.ld file, extracted some section of interest (ignoring userspace) and tried to make an image of what an XIP=y and XIP=n image might look like in the physical memory regions:

cortex-m-linker

With config XIP=y, everything outside of FLASH is either NOLOAD (green) or VMA (virtual memory address) (purple).

With XIP=n:

- `ITCM (LMA)`, `DTCM DATA (LMA)`, `Sections snippets` and `.last_section` are "LOAD" sections that can occur after `BSS`/`NOINIT` "NOLOAD" sections
- `.last_section` "LOAD" section is after `DT_SECTIONS` "NOLOAD" sections
- ITCM and DTCM still use VMA/LMA but the code to copy from LMA to VMA (kernel/xip.c `void z_data_copy(void)`) is excluded if XIP=n (seems like a separate bug)

---

Flattened memory map of the above for an STM32H7B0:

cortex-m-linker-stm32h7b0

If you create a ".bin" file it will be from the first to last blue box.
i.e. if you used ITCM code relocation with XIP=n you would have a ".bin" file of at least 576MB on this SoC.

I've gone a bit off topic, but hopefully visualising it helps make it clearer what is going on

**Comment by tejlmand** [comment]
> c107827 removed `NOLOAD` from `.last_section`, before that commit it was not included in the binary

Correct, but as you have also noticed, the `ITCM`, `DTCM` and the inclusion of linker `section-snippets` also comes after `bss`.
But as those are not included by every board / SoC then removing the `NOLOAD` for the `.last_section` we went from this happening in certain cases to always happen when `XIP=n`.

But I would still argue that the original cause is that `.last_section` was not carefully enough placed in linker.ld by the original commit which introduce it.

> ITCM and DTCM still use VMA/LMA but the code to copy from LMA to VMA (kernel/xip.c void z_data_copy(void)) is excluded if XIP=n (seems like a separate bug)

yep, or perhaps more a feature which has not been taken all the way , that is XIP=n was not considered when doing the `ITCM` / `DTCM` support.

**Review by tejlmand — COMMENTED** [review]
Thanks for both finding, investigating and providing a fix for this issue :+1: .

I have found no side-effects so changes lgtm, but I have a couple additional wishes before a +1:
- can you please take userspace support into consideration in this fix.
  `CONFIG_XIP=n CONFIG_USERSPACE=y` still suffers this problem after this PR, and it's not nice just waiting for someone else hitting this issue again later.
   https://github.com/zephyrproject-rtos/zephyr/blob/7773fe8dd60e9f86760c994090304a0577a1ad2f/include/zephyr/arch/arm/cortex_m/scripts/linker.ld#L283-L292
- Can you please take a look at similar linker.ld files.
   Seems like the `arm64`, `cortex_a_r` suffers the same issue and for the above reason but also because it's nice to have the linker.ld files somewhat aligned in how they look, then it would be nice to have those aligned as well.
   Regarding the SoC specific linker.ld files, then but feel free to ping the SoC maintainers and let them evaluate if the same changes should be done there.

**Comment by gramsay0** [comment]
@tejlmand thanks! I will check userspace and other linker files soon when I find time

**Comment by gramsay0** [comment]
Small update for Cortex-A/R. There was an attempt to move bss/noinit toward the end of the linker file https://github.com/zephyrproject-rtos/zephyr/pull/60368. This ended up getting reverted due to CI failures. The CI failures were later discovered to be unrelated to this change (there was some undefined behavior in one of the tests that just happened to start failing). This PR was never re-applied

Interestingly, Cortex-A/R does not put bss/noinit in different places depending on `CONFIG_USERSPACE` (as Cortex-M does). I'll try to catch up with the original reasons for this https://github.com/zephyrproject-rtos/zephyr/pull/21747

**Comment by tejlmand** [comment]
@gramsay0 I consider updating / aligning with cortex_a_r ld template scripts addressed by @Dat-NguyenDuy in #79502

**Comment by Dat-NguyenDuy** [comment]
> Small update for Cortex-A/R. There was an attempt to move bss/noinit toward the end of the linker file #60368. This ended up getting reverted due to CI failures. The CI failures were later discovered to be unrelated to this change (there was some undefined behavior in one of the tests that just happened to start failing). This PR was never re-applied
> 
> Interestingly, Cortex-A/R does not put bss/noinit in different places depending on `CONFIG_USERSPACE` (as Cortex-M does). I'll try to catch up with the original reasons for this #21747

There was a discussion regarding this topic for cortex-M, but the issue was closed without any actions. I have just re-opened it https://github.com/zephyrproject-rtos/zephyr/issues/24619

**Comment by Dat-NguyenDuy** [comment]
Hi @gramsay0, have you had time to look into the USERSPACE issue on cortex-M? I have tried with your commits and always put bss/noinit sections to the end of the linker script. Run tests\kernel, tests\lib on an cortex-M7 based boards, i see no issue.
Look like the problem describe here: https://github.com/zephyrproject-rtos/zephyr/pull/21747#issuecomment-595768627 is no longer exist.

**Comment by gramsay0** [comment]
@Dat-NguyenDuy yes I had done some testing and also found no issues.

I think this patch fixed the need for bss to be before the kobject_data section with userspace enabled: https://github.com/zephyrproject-rtos/zephyr/pull/33687

See https://github.com/zephyrproject-rtos/zephyr/pull/21747#issuecomment-595768627 (which I believe is what the above patch does):

> The appropriate method may be to reserve space for kobject_data during prelink, but the related process is relatively unfamiliar to me, so it is difficult to modify.

@tejlmand do you agree, and if so should I make that change here in in a follow-up PR?

**Comment by Dat-NguyenDuy** [comment]
Hi @tejlmand and @gramsay0, i'm not sure if there is still work to be done for this PR since i don't see an approve from @tejlmand.

**Review by tejlmand — APPROVED** [review]

**Comment by kartben** [comment]
@ithinuel could you please review and potentially be the second +1 here?

**Review by stephanosio — APPROVED** [review]

**Pull request merged**
